### ISSUE TITLE
Feature Request: editor on materialmaker.org

### ISSUE THREAD

**Issue by therealbnut** [RodZill4/material-maker#611]
I think it’d be really cool to host Material Maker as an editor for shaders on https://materialmaker.org, inspired by sites like https://shadertoy.com. It’d be great for exploring and sharing shaders, and growing the user base.

A few other sites exist like https://godotshaders.com and https://shadered.org/shaders?language=godot but none that really integrate an editor like Material Maker and a live preview.

As far as I know there isn’t any existing site like shadertoy with a visual editor (more accessible to artists), godot or not, and a godot being first would be great!

Implementation-wise I suppose you’d need a cut down version without the rest of the editor, a way to load models, or pick from presets, and a way to communicate the serialized/deserialized data to the browser.

The standalone part could integrate into the existing https://www.materialmaker.org/materials website. Although the existing site would probably need some kind of versioning and forking controls as well.

I’m happy to give it a go with some guidance, although as far as I can tell the site isn’t open source. It’d be cool for the site to be open sourced too, to enable projects like this, and other ergonomic improvements.

**Comment by RodZill4** [comment]
This is absolutely not in the scope of Material Maker. There is a web version, though, but it's pretty limited. https://www.materialmaker.org/public/mm_web/

**Comment by therealbnut** [comment]
@RodZill4 
> not on my roadmap
Yeah I can understand that, I guess what I was wondering is whether you’d be open to it if others did the work? Would you be open to open sourcing the materialmaker.org site to allow contributions like this?

**Comment by RodZill4** [comment]
Accepting contributions on the website would require that I carefully review all contributions, and it's very likely I won't have time for this anytime soon. This is why the website is not opensource (yet).
Say I change my mind, before I can opensource the website, I'd have to cleanup and refactor the code, and remove all pre-1.3 code (the way MM communicates with the website changed in 1.3). This will not happen before 1.5 is available (I won't have time before, plus I hate webdev :D ).

**Comment by therealbnut** [comment]
Thanks, I’m not a web dev fan either - it’d be nice if you could find another owner more into that side of things.

Thank you for the info, and good luck working toward 1.5! I’d appreciate if you let me know if anything changes on this issue, I’d be happy to help implement it.

**Comment by RodZill4** [comment]
Maybe we could chat here about what you have in mind.
At some point I wanted to port the shader generation and renderer to Javascript so materials could be configured, rendered and shown in 3D on the website. But this would be a lot of work (especially keeping it up to date) unless you have a GdScript to Javascript translator. ;)
Editing could have been a second step (materials are saved as fairly simple JSON), nodes are easy to understand etc.
But with Godot 4 I plan to focus on more painting and texturing objects, with heavy usage of compute shaders, so a web version will probably be less and less possible.

**Comment by therealbnut** [comment]
Have you considered instead of JavaScript, port the gdscript to a C++ (or another language) godot plug-in? Then it could be cross-compiled to WebAssembly and used directly in the browser with the same code.

Although I was expecting you’d be using Godot’s built-in support for web export to avoid any porting. The whole editor is made in-engine, so I expected a cut-down version could be done without any porting, assuming MM can factor to be separated from the editor. Although I thought that’s what you’re already doing with mm_web.

I expect Godot to ensure its compute shaders can be compatible with web, for its web export. There is web compatible technology for everything I expect you’d need. Although I’m not completely sure if texture painting is compatible with shareable reusable materials, as a workflow. In my mind texture painting would have been disabled for the site, to focus on shareable materials. If it’s not enabled then compatibility doesn’t matter… but I’ll have to give that workflow some more thought.

If you wanted the whole thing independent of the godot engine, and compatible with web via cross-compilation, I know what technology choices I would make - and I am happy to discuss that. Although it _is_ a much larger change than I was imagining. Reusing godot’s UI and web compatibility makes it a lot easier.

**Comment by Zireael07** [comment]
https://codeberg.org/LinuxUserGD/gdscript-transpiler-bin is an existing gdscript to C++ transpiler. Note how many hoops it jumps through. If you want a compile to WASM language you probably should write your own fork of GDScript tbh.

Godot's web support suffers from many warts unfortunately.

**Comment by therealbnut** [comment]
I was assuming it’d use the godot-web’s gdscript interpreter, not trying to compile it. If you want/need to compile it, then I think it’d be better to port the scripts to C++ or similar that can be more easily cross-compiled.

**Comment by Zireael07** [comment]
I was referring to this 

> Have you considered instead of JavaScript, port the gdscript to a C++ (or another language) godot plug-in? Then it could be cross-compiled to WebAssembly and used directly in the browser with the same code.

which was pretty clearly referring to cross-compiling.

Godot's web exports already run gdscript's interpreter without you needing to do anything

**Comment by therealbnut** [comment]
> which was pretty clearly referring to cross-compiling.

Ah yes, I can see how that could have been confusing. I was referring to cross-compiling, but from C++, not from GDScript.

I should have quoted, but that part was replying to the idea of porting the GDScript to JavaScript:

> At some point I wanted to port the shader generation and renderer to Javascript

**Comment by RodZill4** [comment]
Maybe we want to define "what" before chatting about "how"?
